### PR TITLE
[Bug] Fix condition task will cause workflow instance failed

### LINKED ISSUE #16150: [Bug] Workflow status is success when task execute failed
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

I have a workflow like below, `pass` node is a forbidden node.

After execute, the workflow instance status is success.

image

This is caused by we didn't put the failed node into error map

### What you expected to happen

xx

### How to reproduce

xx

### Anything else

_No response_

### Version

3.2.x

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by ruanwenjun**
## Purpose of the pull request

close https://github.com/apache/dolphinscheduler/issues/16150

## Brief change log

- Put task into error map once a task is failure
- Refactor the condition task check logic.

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java [hunk 1]
```diff
@@ -2015,6 +2013,24 @@ private void saveCacheTaskInstance(TaskInstance taskInstance) {
         }
     }
 
+    /**
+     * Whether the task instance need to put into {@link #errorTaskMap}.
+     * Only the task instance is failed or killed, and it is parent of condition task.
+     * Then it should be put into {@link #errorTaskMap}.
+     * <p> Once a task instance is put into {@link #errorTaskMap}, it will be thought as failed and make the workflow be failed.
+     */
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java:2021]: :+1:

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/condition/ConditionLogicTask.java [hunk 2]
```diff
@@ -40,37 +38,34 @@
...
         // calculate the conditionResult
         DependResult conditionResult = calculateConditionResult();
-        TaskExecutionStatus taskExecutionStatus =
-                (conditionResult == DependResult.SUCCESS) ? TaskExecutionStatus.SUCCESS : TaskExecutionStatus.FAILURE;
-        log.info("The condition result is {}, task instance statue will be: {}", conditionResult, taskExecutionStatus);
-        taskExecutionContext.setCurrentExecutionStatus(taskExecutionStatus);
+        log.info("The condition result is {}", conditionResult);
+        taskParameters.setConditionSuccess(conditionResult == DependResult.SUCCESS);
+        taskInstance.setConditionsParameters(taskParameters);
+        taskExecutionContext.setCurrentExecutionStatus(TaskExecutionStatus.SUCCESS);
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/condition/ConditionLogicTask.java:68]: Need to update the docs too.

see https://dolphinscheduler.apache.org/zh-cn/docs/3.2.1/guide/task/conditions

File: dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/utils/DagHelperTest.java [hunk 3]
```diff
@@ -291,43 +298,13 @@ public void testConditionPostNode() throws IOException {
...
-        // dag: 1-2-3-5-7 4-3-6
-        // 3-if , complete:1/2/3/4
-        // 1.failure:3 expect post:6 skip:5/7
-        dag = generateDag2();
-        skipNodeList.clear();
-        completeTaskList.clear();
-        taskInstance.setSwitchDependency(getSwitchNode());
-        completeTaskList.put(1L, taskInstance);
-        postNodes = DagHelper.parsePostNodes(1L, skipNodeList, dag, completeTaskList);
-        Assertions.assertEquals(1, postNodes.size());
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/utils/DagHelperTest.java:330]: Maybe should keep the switch UT?

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/condition/ConditionLogicTask.java (hunk 2 above)
Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/condition/ConditionLogicTask.java:68]: I find the doc is correct, we don't need to update it.

File: dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/utils/DagHelperTest.java (hunk 3 above)
Comment (ruanwenjun) [reply dolphinscheduler-service/src/test/java/org/apache/dolphinscheduler/service/utils/DagHelperTest.java:330]: We have switch UT case, and it's better to split the switch case from condition case. we need to make the case is small enough to maintain.

**Review by SbloodyS — APPROVED** [review]
LGTM.

**Review by rickchengx — APPROVED** [review]
LGTM

**Pull request merged**